[NativeImageBuildStep] Unable to parse GraalVM for JDK 26

We see new native-image integration tests failing because Quarkus isn't able to parse the following JDK 26-based GraalVM version:

```
$ native-image --version
native-image 26 2026-03-17
GraalVM Runtime Environment GraalVM CE 26-dev+1.1 (build 26+1-jvmci-b01)
Substrate VM GraalVM CE 26-dev+1.1 (build 26+1, serial gc)
```

This results in the `picocli-native` integration test to fail with an error along the lines of this:

```
[INFO] --- quarkus:999-SNAPSHOT:track-config-changes (track-config-changes) @ quarkus-integration-test-picocli-native ---
[INFO] 
[INFO] --- compiler:3.14.0:compile (default-compile) @ quarkus-integration-test-picocli-native ---
[INFO] Recompiling the module because of changed source code.
[INFO] Compiling 21 source files with javac [debug deprecation parameters release 17] to target/classes
[INFO] 
[INFO] --- resources:3.3.1:testResources (default-testResources) @ quarkus-integration-test-picocli-native ---
[INFO] skip non existing resourceDirectory /home/sgehwolf/Documents/openjdk/quarkus/quarkus-source/integration-tests/picocli-native/src/test/resources
[INFO] 
[INFO] --- compiler:3.14.0:testCompile (default-testCompile) @ quarkus-integration-test-picocli-native ---
[INFO] Recompiling the module because of changed dependency.
[INFO] Compiling 4 source files with javac [debug deprecation parameters release 17] to target/test-classes
[INFO] 
[INFO] --- surefire:3.5.3:test (default-test) @ quarkus-integration-test-picocli-native ---
[INFO] Skipped
[INFO] 
[INFO] --- quarkus:999-SNAPSHOT:build (default) @ quarkus-integration-test-picocli-native ---
[INFO] [quarkus-build-caching-extension] Quarkus previous configuration not found
[INFO] [quarkus-build-caching-extension] Quarkus build goal marked as not cacheable
[INFO] [quarkus-build-caching-extension] Quarkus previous configuration not found
[INFO] [quarkus-build-caching-extension] Quarkus build goal marked as not cacheable
[INFO] [io.quarkus.deployment.pkg.steps.JarResultBuildStep] Building native image source jar: /home/sgehwolf/Documents/openjdk/quarkus/quarkus-source/integration-tests/picocli-native/target/quarkus-integration-test-picocli-native-999-SNAPSHOT-native-image-source-jar/quarkus-integration-test-picocli-native-999-SNAPSHOT-runner.jar
[INFO] [io.quarkus.deployment.pkg.steps.NativeImageBuildStep] Building native image from /home/sgehwolf/Documents/openjdk/quarkus/quarkus-source/integration-tests/picocli-native/target/quarkus-integration-test-picocli-native-999-SNAPSHOT-native-image-source-jar/quarkus-integration-test-picocli-native-999-SNAPSHOT-runner.jar
[INFO] ------------------------------------------------------------------------
[INFO] BUILD FAILURE
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  4.801 s
[INFO] Finished at: 2025-06-16T12:09:32+02:00
[INFO] ------------------------------------------------------------------------
[INFO] Njord session closed
[INFO] 14 goals, 14 executed
[ERROR] Failed to execute goal io.quarkus:quarkus-maven-plugin:999-SNAPSHOT:build (default) on project quarkus-integration-test-picocli-native: Failed to build quarkus application: io.quarkus.builder.BuildException: Build failure: Build failed due to errors
[ERROR] 	[error]: Build step io.quarkus.deployment.pkg.steps.NativeImageBuildStep#build threw an exception: java.lang.RuntimeException: Failed to get GraalVM version
[ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildRunner.getGraalVMVersion(NativeImageBuildRunner.java:41)
[ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildStep.build(NativeImageBuildStep.java:246)
[ERROR] 	at java.base/java.lang.invoke.MethodHandle.invokeWithArguments(MethodHandle.java:735)
[ERROR] 	at io.quarkus.deployment.ExtensionLoader$3.execute(ExtensionLoader.java:856)
[ERROR] 	at io.quarkus.builder.BuildContext.run(BuildContext.java:255)
[ERROR] 	at org.jboss.threads.ContextHandler$1.runWith(ContextHandler.java:18)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.doRunWith(EnhancedQueueExecutor.java:2651)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$Task.run(EnhancedQueueExecutor.java:2630)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor.runThreadBody(EnhancedQueueExecutor.java:1622)
[ERROR] 	at org.jboss.threads.EnhancedQueueExecutor$ThreadBody.run(EnhancedQueueExecutor.java:1589)
[ERROR] 	at java.base/java.lang.Thread.run(Thread.java:1474)
[ERROR] 	at org.jboss.threads.JBossThread.run(JBossThread.java:501)
[ERROR] Caused by: java.lang.NumberFormatException: For input string: "null"
[ERROR] 	at java.base/java.lang.NumberFormatException.forInputString(NumberFormatException.java:67)
[ERROR] 	at java.base/java.lang.Integer.parseInt(Integer.java:565)
[ERROR] 	at java.base/java.lang.Integer.parseInt(Integer.java:662)
[ERROR] 	at java.base/java.util.stream.ReferencePipeline$4$1.accept(ReferencePipeline.java:231)
[ERROR] 	at java.base/java.util.Spliterators$ArraySpliterator.forEachRemaining(Spliterators.java:1024)
[ERROR] 	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:570)
[ERROR] 	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:560)
[ERROR] 	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:635)
[ERROR] 	at java.base/java.util.stream.AbstractPipeline.evaluateToArrayNode(AbstractPipeline.java:291)
[ERROR] 	at java.base/java.util.stream.IntPipeline.toArray(IntPipeline.java:564)
[ERROR] 	at io.quarkus.runtime.graal.GraalVM$Version.breakdownVersion(GraalVM.java:127)
[ERROR] 	at io.quarkus.runtime.graal.GraalVM$Version.<init>(GraalVM.java:116)
[ERROR] 	at io.quarkus.deployment.pkg.steps.GraalVM$Version.<init>(GraalVM.java:229)
[ERROR] 	at io.quarkus.deployment.pkg.steps.GraalVM$VersionParseHelper.parse(GraalVM.java:89)
[ERROR] 	at io.quarkus.deployment.pkg.steps.GraalVM$Version.of(GraalVM.java:279)
[ERROR] 	at io.quarkus.deployment.pkg.steps.NativeImageBuildRunner.getGraalVMVersion(NativeImageBuildRunner.java:38)
[ERROR] 	... 11 more
[ERROR] -> [Help 1]
[ERROR] 
[ERROR] To see the full stack trace of the errors, re-run Maven with the -e switch.
[ERROR] Re-run Maven using the -X switch to enable full debug logging.
[ERROR] 
[ERROR] For more information about the errors and possible solutions, please read the following articles:
[ERROR] [Help 1] http://cwiki.apache.org/confluence/display/MAVEN/MojoExecutionException
```

We also see similar failures in our Mandrel CI here:
https://github.com/graalvm/mandrel/actions/runs/15646482173/job/44085716121#step:12:455

--- a/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/GraalVM.java
+++ b/core/deployment/src/main/java/io/quarkus/deployment/pkg/steps/GraalVM.java
@@ -28,7 +28,6 @@ public final class GraalVM {
         private static final String OPT = "(?:-(?<OPT>[-a-zA-Z0-9.]+))?";
         private static final String VSTR_FORMAT = VNUM + PRE + BUILD + OPT;
 
-        private static final String VNUM_GROUP = "VNUM";
         private static final String VENDOR_VERSION_GROUP = "VENDOR";
         private static final String BUILD_INFO_GROUP = "BUILDINFO";
 
@@ -67,7 +66,7 @@ public final class GraalVM {
 
                 String vendorVersion = secondMatcher.group(VENDOR_VERSION_GROUP);
 
-                String graalVersion = graalVersion(javaVersion, v.feature());
+                String graalVersion = graalVersion(javaVersion, v);
                 if (vendorVersion.contains("-dev")) {
                     graalVersion = graalVersion + "-dev";
                 }
@@ -137,7 +136,7 @@ public final class GraalVM {
             return null;
         }
 
-        private static String graalVersion(String buildInfo, int jdkFeature) {
+        private static String graalVersion(String buildInfo, Runtime.Version v) {
             if (buildInfo == null) {
                 return null;
             }
@@ -152,7 +151,10 @@ public final class GraalVM {
             if (versMatcher.find()) {
                 return matchVersion(version);
             } else {
-                return Version.GRAAL_MAPPING.get(Integer.toString(jdkFeature));
+                // Only versions from JDK 22 to JDK 25 had GraalVM version mappings.
+                // Use the JDK version triplet for JDK N where N > 25.
+                String fullJDKVersion = String.format("%d.%d.%d", v.feature(), v.interim(), v.update());
+                return Version.GRAAL_MAPPING.getOrDefault(Integer.toString(v.feature()), fullJDKVersion);
             }
         }
 

--- a/core/deployment/src/test/java/io/quarkus/deployment/pkg/steps/GraalVMTest.java
+++ b/core/deployment/src/test/java/io/quarkus/deployment/pkg/steps/GraalVMTest.java
@@ -7,6 +7,7 @@ import static org.assertj.core.api.Assertions.assertThat;
 
 import java.util.stream.Stream;
 
+import org.junit.jupiter.api.AfterAll;
 import org.junit.jupiter.api.Assertions;
 import org.junit.jupiter.api.Test;
 import org.junit.jupiter.params.ParameterizedTest;
@@ -18,8 +19,21 @@ import io.quarkus.runtime.graal.GraalVM.Distribution;
 
 public class GraalVMTest {
 
+    public static final String GRAALVM_VENDOR_VERSION_PROP = "org.graalvm.vendorversion";
+
+    @AfterAll
+    public static void tearDownAfterAll() {
+        System.clearProperty(GRAALVM_VENDOR_VERSION_PROP);
+    }
+
     @Test
     public void testGraalVMVersionDetected() {
+        // Version detection after JDK 25 which no longer uses a GraalVM version mapping,
+        // but uses the JDK version instead
+        assertVersion(new Version("GraalVM 26.0.0", "26.0.0", GRAALVM), GRAALVM,
+                Version.of(Stream.of(("native-image 26 2026-03-17\n"
+                        + "GraalVM Runtime Environment GraalVM CE 26-dev+1.1 (build 26+1-jvmci-b01)\n"
+                        + "Substrate VM GraalVM CE 26-dev+1.1 (build 26+1, serial gc)").split("\\n"))));
         // Version detection after: https://github.com/oracle/graal/pull/6302 (3 lines of version output)
         assertVersion(new Version("GraalVM 23.0.0", "23.0.0", GRAALVM), MANDREL,
                 Version.of(Stream.of(("native-image 17.0.6 2023-01-17\n"
@@ -169,6 +183,19 @@ public class GraalVMTest {
     }
 
     @Test
+    public void testGraalVM26CommunityVersionParser() {
+        final Version version = Version.of(Stream.of(("native-image 26 2026-03-17\n"
+                + "GraalVM Runtime Environment GraalVM CE 26-dev+1.1 (build 26+1-jvmci-b01)\n"
+                + "Substrate VM GraalVM CE 26-dev+1.1 (build 26+1, serial gc)").split("\\n")));
+        assertThat(version.toString().contains(GRAALVM.name()));
+        assertThat(version.getVersionAsString()).isEqualTo("26.0.0-dev");
+        assertThat(version.javaVersion.toString()).isEqualTo("26+1-jvmci-b01");
+        assertThat(version.javaVersion.feature()).isEqualTo(26);
+        assertThat(version.javaVersion.interim()).isEqualTo(0);
+        assertThat(version.javaVersion.update()).isEqualTo(0);
+    }
+
+    @Test
     public void testGraalVMVersionsOlderThan() {
         assertOlderThan("native-image 21 2023-09-19\n" +
                 "GraalVM Runtime Environment GraalVM CE 21+35.1 (build 21+35-jvmci-23.1-b15)\n" +
@@ -182,6 +209,40 @@ public class GraalVMTest {
                 "native-image 21 2023-09-19\n" +
                         "GraalVM Runtime Environment GraalVM CE 21+35.1 (build 21+35-jvmci-23.1-b15)\n" +
                         "Substrate VM GraalVM CE 21+35.1 (build 21+35, serial gc)\n");
+    }
+
+    /*
+     * Exercise the code path used at native-image build time where the org.graalvm.vendorversion
+     * property is being fed to the GraalVM version parsing machinery.
+     */
+    @ParameterizedTest
+    // @formatter:off
+    @ValueSource(strings = {
+    // GraalVM CE/Community
+    "GraalVM CE 26-dev+1.1 |26.0|26|0",
+    "GraalVM CE 25-dev+26.1|25.0|25|0",
+    "GraalVM CE 24.0.1+9.1 |24.2|24|1",
+    // Mandrel
+    "Mandrel-24.2.1.0-Final|24.2|24|1",
+    "Mandrel-23.1.7.0-1b2  |23.1|21|7",
+    // Liberica-NIK
+    "Liberica-NIK-23.1.7-1 |23.1|21|7",
+    "Liberica-NIK-24.2.1-1 |24.2|24|1",
+    })
+    // @formatter:on
+    public void testGraalVMRuntimeVersion(String param) {
+        String tokens[] = param.split("\\|", 4);
+        Assertions.assertTrue(tokens.length == 4);
+        String propertyVal = tokens[0].trim();
+        String expectedMajorMinor = tokens[1].trim();
+        int expectedJDKFeature = Integer.parseInt(tokens[2].trim());
+        int expectedJDKUpdate = Integer.parseInt(tokens[3].trim());
+
+        System.setProperty(GRAALVM_VENDOR_VERSION_PROP, propertyVal);
+        io.quarkus.runtime.graal.GraalVM.Version v = io.quarkus.runtime.graal.GraalVM.Version.getCurrent();
+        Assertions.assertEquals(expectedMajorMinor, v.getMajorMinorAsString());
+        Assertions.assertEquals(expectedJDKFeature, v.javaVersion.feature());
+        Assertions.assertEquals(expectedJDKUpdate, v.javaVersion.update());
     }
 
     /**

--- a/core/runtime/src/main/java/io/quarkus/runtime/graal/GraalVM.java
+++ b/core/runtime/src/main/java/io/quarkus/runtime/graal/GraalVM.java
@@ -1,7 +1,10 @@
 package io.quarkus.runtime.graal;
 
 import java.util.Arrays;
+import java.util.Collections;
+import java.util.HashMap;
 import java.util.Map;
+import java.util.Map.Entry;
 import java.util.regex.Matcher;
 import java.util.regex.Pattern;
 import java.util.stream.Collectors;
@@ -40,16 +43,23 @@ public final class GraalVM {
 
         private static final String VERSION_GROUP = "VNUM";
 
-        private static final Version UNKNOWN_VERSION = null;
-
         static Version parse(String value) {
             Matcher versionMatcher = VENDOR_VERS_PATTERN.matcher(value);
             if (versionMatcher.find()) {
                 String vendor = versionMatcher.group(VENDOR_PREFIX_GROUP);
                 if (GRAALVM_CE_VERS_PREFIX.equals(vendor) || ORACLE_GRAALVM_VERS_PREFIX.equals(vendor)) {
                     String version = versionMatcher.group(VERSION_GROUP);
-                    String jdkFeature = version.split("\\.", 2)[0];
-                    return new Version(value, Version.GRAAL_MAPPING.get(jdkFeature), Distribution.GRAALVM);
+                    String tokens[] = version.split("\\.", 3);
+                    String jdkFeature = tokens[0];
+                    String jdkVers = jdkFeature;
+                    if (tokens.length == 3) {
+                        String interim = tokens[1];
+                        String update = tokens[2].split("\\+")[0];
+                        jdkVers = String.format("%s.%s.%s", jdkFeature, interim, update);
+                    }
+                    // For JDK 26+ there is no more version mapping use the JDK version
+                    String versionMapping = Version.GRAAL_MAPPING.getOrDefault(jdkFeature, version);
+                    return new Version(value, versionMapping, jdkVers, Distribution.GRAALVM);
                 } else if (LIBERICA_NIK_VERS_PREFIX.equals(vendor)) {
                     return new Version(value, versionMatcher.group(VERSION_GROUP), Distribution.LIBERICA);
                 } else if (MANDREL_VERS_PREFIX.equals(vendor)) {
@@ -78,8 +88,18 @@ public final class GraalVM {
                 "21", "23.1",
                 "22", "24.0",
                 "23", "24.1",
-                "24", "24.2",
-                "25", "25.0");
+                "24", "24.2");
+        // Mapping of community major.minor pair to the JDK major version based on
+        // GRAAL_MAPPING
+        private static final Map<String, String> MANDREL_JDK_REV_MAP;
+
+        static {
+            Map<String, String> reverseMap = new HashMap<>(GRAAL_MAPPING.size());
+            for (Entry<String, String> entry : GRAAL_MAPPING.entrySet()) {
+                reverseMap.put(entry.getValue(), entry.getKey());
+            }
+            MANDREL_JDK_REV_MAP = Collections.unmodifiableMap(reverseMap);
+        }
 
         /**
          * The minimum version of GraalVM supported by Quarkus.
@@ -97,6 +117,7 @@ public final class GraalVM {
          */
         public static final Version MINIMUM_SUPPORTED = CURRENT;
 
+        private static final String DEFAULT_JDK_VERSION = "21";
         protected final String fullVersion;
         public final Runtime.Version javaVersion;
         protected final Distribution distribution;
@@ -104,7 +125,10 @@ public final class GraalVM {
         private String suffix;
 
         Version(String fullVersion, String version, Distribution distro) {
-            this(fullVersion, version, "21", distro);
+            this(fullVersion, version,
+                    distro == Distribution.MANDREL || distro == Distribution.LIBERICA ? communityJDKvers(version)
+                            : DEFAULT_JDK_VERSION,
+                    distro);
         }
 
         Version(String fullVersion, String version, String javaVersion, Distribution distro) {
@@ -125,6 +149,33 @@ public final class GraalVM {
                 version = version.substring(0, dash);
             }
             this.versions = Arrays.stream(version.split("\\.")).mapToInt(Integer::parseInt).toArray();
+        }
+
+        /*
+         * Reconstruct the JDK version from the given GraalVM community version (Mandrel or Liberica)
+         */
+        private static String communityJDKvers(String communityVersion) {
+            try {
+                String[] parts = communityVersion.split("\\.", 4);
+                int major = Integer.parseInt(parts[0]);
+                int minor = Integer.parseInt(parts[1]);
+                if ((major == 23 && minor > 0) ||
+                        major > 23) {
+                    String mandrelMajorMinor = String.format("%s.%s", parts[0], parts[1]);
+                    // If we don't find a reverse mapping we use a JDK version >= 25, thus
+                    // the feature version is the first part of the quadruple.
+                    String feature = MANDREL_JDK_REV_MAP.getOrDefault(mandrelMajorMinor, parts[0]);
+                    // Heuristic: The update version of Mandrel and the JDK match.
+                    // Interim is usually 0 for the JDK version.
+                    return String.format("%s.%s.%s", feature, "0", parts[2]);
+                }
+            } catch (Throwable e) {
+                // fall-through do default
+                Log.warnf("Failed to parse JDK version from GraalVM version: %s. Defaulting to currently supported version %s ",
+                        communityVersion,
+                        DEFAULT_JDK_VERSION);
+            }
+            return DEFAULT_JDK_VERSION;
         }
 
         @Override

